### PR TITLE
Ban querySelector and querySelectorAll

### PR DESCRIPTION
**Pull request by Josehower**
As part of this PR I created this issue where I am asking for an official response about the topic https://github.com/reactjs/reactjs.org/issues/4626

### CODE REVIEW

File: index.cjs [hunk 1]
```diff
@@ -171,6 +171,18 @@ module.exports = {
         message:
           'Update content and elements with React instead of using location.reload()',
       },
+      {
+        selector:
+          "ExpressionStatement CallExpression[callee.object.name='document'][callee.property.name='querySelector']",
+        message:
+          'Use Controlled components instead of document.querySelector()',
```
[the remark this answers is not in the document]
Comment (karlhorky) [reply index.cjs:178]: also look at the other messages, see if there's anything you can take from them

**Review by karlhorky — CHANGES_REQUESTED** [review]
Thanks for this! Some small things above

File: index.cjs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Josehower) [reply index.cjs]: https://github.com/upleveled/eslint-config-upleveled/pull/105/files

[the remark this answers is not in the document]
Comment (Josehower) [reply index.cjs]: https://github.com/upleveled/eslint-config-upleveled/pull/105/files

[the remark this answers is not in the document]
Comment (Josehower) [reply index.cjs]: https://github.com/upleveled/eslint-config-upleveled/pull/105/commits/7728c9a8c01039232fb4ecfa8357354dcee5dbd4

File: index.cjs [hunk 2]
```diff
@@ -165,6 +165,24 @@ module.exports = {
     // Warn on restricted syntax
     'no-restricted-syntax': [
       'warn',
+      {
+        selector:
+          "ExpressionStatement CallExpression[callee.object.name='document'][callee.property.name='querySelector']",
+        message:
+          'Using `document.querySelector()` can give unexpected results in between renders or if there are multiple elements that match the query (for example, when the component is used more than once), [learn more](https://github.com/reactjs/reactjs.org/issues/4626#issuecomment-1117535930)',
```
Comment (karlhorky) [inline index.cjs:172]: are we allowed to use Markdown links in ESLint messages?

Comment (Josehower) [reply index.cjs:172]: lol I am not sure ill check 
but I think so.

Comment (Josehower) [reply index.cjs]: [273c877](https://github.com/upleveled/eslint-config-upleveled/pull/105/commits/273c877c5bfa3581e21d34e9436fdd045fd64af3)

File: index.cjs [hunk 3]
```diff
@@ -165,6 +165,24 @@ module.exports = {
     // Warn on restricted syntax
     'no-restricted-syntax': [
       'warn',
+      {
+        selector:
+          "ExpressionStatement CallExpression[callee.object.name='document'][callee.property.name='querySelector']",
+        message: `Using document.querySelectorAll() can give unexpected results specially if there are multiple elements that match the query (for example, when the component is used more than once), as an advise use State or Refs instead
+        https://github.com/reactjs/reactjs.org/issues/4626#issuecomment-1117535930`,
```
Comment (karlhorky) [inline index.cjs:172]: maybe the lines after the newlines should be at the beginning of the line? eg. with no indentation?

File: index.cjs [hunk 4]
```diff
@@ -165,6 +165,24 @@ module.exports = {
...
+        selector:
+          "ExpressionStatement CallExpression[callee.object.name='document'][callee.property.name='querySelector']",
+        message: `Using document.querySelector() can give unexpected results, especially if multiple elements match the query (for example, when the component is used more than once); as a piece of advice, use State or Refs instead
+        https://github.com/reactjs/reactjs.org/issues/4626#issuecomment-1117535930`,
+      },
+      {
+        selector:
+          "ExpressionStatement CallExpression[callee.object.name='document'][callee.property.name='querySelectorAll']",
+        message: `Using document.querySelectorAll() can give unexpected results, especially if multiple elements match the query (for example, when the component is used more than once); as a piece of advice, use State or Refs instead
+        https://github.com/reactjs/reactjs.org/issues/4626#issuecomment-1117535930`,
```
Comment (karlhorky) [inline index.cjs:178]: Can you add a comment about why we can't use Markdown links yet (with all the links to the open issues in VS Code / VS Code extension)?

Here are some issues that we both found, not sure if they are all relevant (or if there are more relevant links which are not in this list):

- https://github.com/microsoft/vscode/issues/11847
- https://github.com/microsoft/vscode-eslint/issues/418
- https://github.com/microsoft/vscode/issues/54272

Comment (Josehower) [reply index.cjs]: https://github.com/upleveled/eslint-config-upleveled/pull/105/commits/085d47f972cff939f57da74420ddcd2e8f55f758

Comment (Josehower) [reply index.cjs]: https://github.com/upleveled/eslint-config-upleveled/issues/123

File: index.cjs [hunk 5]
```diff
@@ -165,6 +165,35 @@ module.exports = {
...
+       * - https://github.com/microsoft/vscode-eslint/issues/418
+       * - https://github.com/microsoft/vscode/issues/11847
+       *
+       * UpLeveled issue with an alternative solution to the problem:
+       * - https://github.com/upleveled/eslint-config-upleveled/issues/123
+       */
+      {
+        selector:
+          "ExpressionStatement CallExpression[callee.object.name='document'][callee.property.name='querySelector']",
+        message: `Using document.querySelector() can give unexpected results, especially if multiple elements match the query (for example, when the component is used more than once); as a piece of advice, use State or Refs instead
```
Comment (karlhorky) [inline index.cjs:181]:
```suggestion
        message: `Using document.querySelector() can lead to problems, and is not commonly used in React code - prefer instead usage of basic React patterns such as state and controlled components
```

File: index.cjs [hunk 6]
```diff
@@ -165,6 +165,35 @@ module.exports = {
...
+      /**
+       * Currently it is not possible to use markdown links inside of the eslint warn/error message.
+       * related links:
+       * - https://github.com/microsoft/vscode/issues/54272
+       * - https://github.com/microsoft/vscode-eslint/issues/418
+       * - https://github.com/microsoft/vscode/issues/11847
+       *
+       * UpLeveled issue with an alternative solution to the problem:
+       * - https://github.com/upleveled/eslint-config-upleveled/issues/123
+       */
```
Comment (karlhorky) [inline index.cjs:177]:
```suggestion
       // Currently it is not possible to use Markdown eg. links in ESLint warnings / error messages
       //
       // FIXME: Switch to a custom rule
       // https://github.com/upleveled/eslint-config-upleveled/issues/123
```

File: index.cjs [hunk 7]
```diff
@@ -165,6 +165,29 @@ module.exports = {
     // Warn on restricted syntax
     'no-restricted-syntax': [
       'warn',
+       // Currently it is not possible to use Markdown eg. links in ESLint warnings / error messages
+       //
+       // FIXME: Switch to a custom rule
+       // https://github.com/upleveled/eslint-config-upleveled/issues/123
```
Comment (karlhorky) [inline index.cjs:171]:
```suggestion
      // Currently it is not possible to use Markdown eg. links in ESLint warnings / error messages
      //
      // FIXME: Switch to a custom rule
      // https://github.com/upleveled/eslint-config-upleveled/issues/123
```

**Review by karlhorky — APPROVED** [review]
LGTM, thanks!

**Pull request merged**